rollup/da_syncer/l1_message_queue_height_finder.go

```diff
@@ -0,0 +1,127 @@
+	calldataBlobSource, err := da.NewCalldataBlobSource(ctx, l1height, l1Reader, blobClient, db)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create calldata blob source: %w", err)
+	}
+
+	return &L1MessageQueueHeightFinder{
+		ctx:                ctx,
+		calldataBlobSource: calldataBlobSource,
+		l1Reader:           l1Reader,
+	}, nil
+}
+
+// FindL1MessageQueueHeight finds the L1 message queue height for the target batch.
+// It does so by:
+// 1. find bundle in which target batch was finalized
+// 2. fetch the tx of the bundle to get the height of the L1 message queue after the bundle
+// 3. with this information we can calculate the L1 message count for each batch from last finalized bundle to the target batch.
+func (f *L1MessageQueueHeightFinder) FindL1MessageQueueHeight(targetBatch uint64) (uint64, error) {
+	batches := make(map[uint64]da.EntryWithBlocks)
+
+	// 1. find bundle in which target batch was finalized
+	daEntries, err := f.calldataBlobSource.NextData()
+	if err != nil {
+		return 0, fmt.Errorf("failed to get next data: %w", err)
+	}
+
+	var finalizedBundle *da.FinalizeBatch
+	for _, daEntry := range daEntries {
+		switch daEntry.Type() {
+		case da.CommitBatchV0Type, da.CommitBatchWithBlobType:
+			daEntryWithBlocks, ok := daEntry.(da.EntryWithBlocks)
+			if !ok {
+				return 0, fmt.Errorf("unexpected type of daEntry: %T, expected EntryWithBlocks", daEntry)
+			}
+
+			// save the batch for later use
+			batches[daEntry.BatchIndex()] = daEntryWithBlocks
+		case da.RevertBatchType:
+			if err = f.handleRevertEvent(batches, daEntry.Event()); err != nil {
+				return 0, fmt.Errorf("failed to handle revert event: %w", err)
+			}
+		case da.FinalizeBatchType:
+			// the finalized event is triggered only for the last batch in the bundle:
+			// we found the bundle in which the target batch was finalized
+			if daEntry.BatchIndex() >= targetBatch {
+				finalizedBundle = daEntry.(*da.FinalizeBatch)
+				break
+			}
+
+		default:
+			return 0, fmt.Errorf("unexpected type of daEntry: %T", daEntry)
+		}
+	}
+
+	if finalizedBundle == nil {
+		return 0, fmt.Errorf("failed to find the bundle in which the target batch was finalized")
+	}
+
+	// 2. fetch tx of the bundle to get the LastProcessedMessageQueueIndex after the bundle and the first batch within the bundle.
+	args, err := f.l1Reader.FetchFinalizeTxDataPostEuclidV2(finalizedBundle.Event().(*l1.FinalizeBatchEvent))
+	if err != nil {
+		return 0, fmt.Errorf("failed to fetch finalize tx data: %w", err)
+	}
+
+	// 3. with this information we can calculate the L1 message queue height for target batch: for each batch from last finalized batch to the target batch subtract L1 messages popped in the batch from L1 message queue height
+	lastBatchInBundle := finalizedBundle.BatchIndex()
+	l1MessageQueueHeight := args.LastProcessedMessageQueueIndex.Uint64()
```

Comment:
Need to update to `totalMessagesPoppedAfter` as per yesterday's change.

Reply:
Code change: https://github.com/scroll-tech/scroll-contracts/pull/74/commits/f56af34b59b9c3a49e80ac09374df324346236a0, discussion: https://scrollco.slack.com/archives/C02PKAQ2430/p1740570172178179?thread_ts=1740567979.601539&cid=C02PKAQ2430

Reply:
done in cef11180bb337a2a6175ad35a9d70346dd09c6cd